### PR TITLE
Add support for Azure authentication using Azure AD Tokens

### PR DESCRIPTION
**Pull request by alaeddine-13**
## Description

Adds support for Azure authentication using Azure AD Tokens

## Pull request checklist
  - [x] Support from environment variable
  - [x] Support from model YAML config
- tests
  - [ ] Do we have tests? Should we add more?
  - [ ] Should we parametrize tests?
  - [ ] Should we add hypothesis tests?
- documentation
  - [ ] add docstrings
  - [ ] if a module is new: add API reference file for Sphinx in `docs/source/api-reference`
  - [ ] check API doc of Sphinx - build and show with `make sphinx && make open-sphinx`
- [x] fix checks
  - start locale checks with `make check`
  - fix formatting with `make format`
- Python type info
  - [x] check and improve Python type info
  - [x] double check if return types are specified - including `-> None`
- [x] check TODO comments in code
- [x] check FIXME comments in code
- [x] check if we want to increase the version or rc number
- [x] check if the copyright needs an update
  - file header
  - README.md
  - LICENSE file

### CODE REVIEW

**Comment by PhilipMay** [comment]
@alaeddine-13 many thanks for thos PR.
Can you please add a copyright line to every file you touched.

Like this:
```
# Copyright (c) 2024 your_name_here, Deutsche Telekom AG
```

**Comment by PhilipMay** [comment]
@alaeddine-13 please use the formatter. can do that with `make format` and `make check`.
That should fix some CI problems.

A very detailed guide is here: https://github.com/telekom/mltb2/blob/main/CONTRIBUTING.md

File: mltb2/openai.py [hunk 1]
```diff
@@ -340,19 +343,47 @@ class OpenAiAzureChat(OpenAiChat):
...
     azure_endpoint: str
+    azure_ad_token: Optional[str]
 
     def __post_init__(self) -> None:
         """Do post init."""
         self.client = AzureOpenAI(
             api_key=self.api_key,
             api_version=self.api_version,
             azure_endpoint=self.azure_endpoint,
+            azure_ad_token_provider=lambda: self.azure_ad_token,
```
[the remark this answers is not in the document]
Comment (PhilipMay) [reply mltb2/openai.py:357]: Can you please try to switch this to "just a token" and not a token provider?

File: mltb2/openai.py [hunk 2]
```diff
@@ -340,19 +343,47 @@ class OpenAiAzureChat(OpenAiChat):
             A common value for this is ``2023-05-15``.
         azure_endpoint: The Azure endpoint.
     """
-
+    api_key: Optional[str]
     api_version: str
     azure_endpoint: str
+    azure_ad_token: Optional[str]
```
Comment (PhilipMay) [inline mltb2/openai.py:349]: As discussed -> please change the order here with `None` init.

File: mltb2/openai.py [hunk 3]
```diff
@@ -340,19 +343,47 @@ class OpenAiAzureChat(OpenAiChat):
...
         )
         self.async_client = AsyncAzureOpenAI(
             api_key=self.api_key,
             api_version=self.api_version,
             azure_endpoint=self.azure_endpoint,
+            azure_ad_token_provider=lambda: self.azure_ad_token,
         )
+
+    @classmethod
+    def from_yaml(cls, yaml_file, **kwargs):
```
Comment (PhilipMay) [inline mltb2/openai.py:367]: As discussed: please use explicit `api_key` and `azure_ad_token` here.

**Comment by PhilipMay** [comment]
@alaeddine-13 please also add your copyright to the README.md and the LICENSE file.

**Comment by PhilipMay** [comment]
Please increase the rc number in `pyproject.toml`.

**Comment by PhilipMay** [comment]
Hi @alaeddine-13 ! To fix the CI issue with ruff: please do a rebase upstream/main. I fixed the issue there.

**Comment by alaeddine-13** [comment]
> Hi @alaeddine-13 ! To fix the CI issue with ruff: please do a rebase upstream/main. I fixed the issue there.

done

**Comment by PhilipMay** [comment]
We have this remaining mypy issue:

```
 Found 3 errors in 1 file (checked 28 source files)
mltb2/openai.py:369: note:          @classmethod
mltb2/openai.py:369: note:          def from_yaml(cls, yaml_file: Any, api_key: Optional[str] = ..., **kwargs: Any) -> Any
mltb2/openai.py:369: note:      Subclass:
mltb2/openai.py:369: note:          @classmethod
mltb2/openai.py:369: note:          def from_yaml(cls, yaml_file: Any, api_key: Optional[str] = ..., azure_ad_token: Optional[str] = ...) -> Any
```

Should we change back to "`kwargs` everywhere" or ignore this issue?

**Comment by PhilipMay** [comment]
@alaeddine-13 mypy now fails with:
```
mltb2/openai.py:364: error: Argument "azure_endpoint" to "AsyncAzureOpenAI" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

The `azure_endpoint` is now optional in your PR. This is the reason for this issue. But AsyncAzureOpenAI expects it.
It is not optional there.

What is the reason to make azure_endpoint optional? Can that be changed back?

**Comment by alaeddine-13** [comment]
@PhilipMay can you run again the CI ? I implemented [this workaround here](https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses) in order to make azure_endpoint a positional argument

**Pull request merged**